[WebGPU] Support tier1 level argument buffers

mwyrzykowski opened the pull request:
#### 8e360957f9200747edec0562115accd8920c744b
<pre>
[WebGPU] Support tier1 level argument buffers
<a href="https://bugs.webkit.org/show_bug.cgi?id=250123">https://bugs.webkit.org/show_bug.cgi?id=250123</a>
&lt;radar://103905011&gt;

Reviewed by Myles C. Maxfield.

Support Tier1 argument buffers which are limited to
pre-A13 devices and some Intel Macs.

Tested on a pre-A13 iPad Pro.

* Source/WebGPU/WebGPU/BindGroup.mm:
(WebGPU::sizeOfEntries):
(WebGPU::Device::createBindGroup):
Support tier1 ABs.

* Source/WebGPU/WebGPU/BindGroupLayout.h:
(WebGPU::BindGroupLayout::create):
Simplify create function.

* Source/WebGPU/WebGPU/BindGroupLayout.mm:
(WebGPU::createArgumentDescriptor):
(WebGPU::addDescriptor):
(WebGPU::Device::createBindGroupLayout):
(WebGPU::BindGroupLayout::BindGroupLayout):
(WebGPU::BindGroupLayout::encodedLength const):
Support tier1 ABs.

* Source/WebGPU/WebGPU/RenderPipeline.mm:
(WebGPU::createEntryFromStructMember):
(WebGPU::RenderPipeline::getBindGroupLayout):
Call Device::createBindLayout instead and pass some additional
information to be compatible with Tier1 ABs.

Canonical link: <a href="https://commits.webkit.org/258651@main">https://commits.webkit.org/258651@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/d188141697254d62bb7f1f1ec796d22688fda73f

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/6/builds/102581 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/77/builds/11712 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/43/builds/35614 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/8/builds/111848 "Built successfully") | [💥 🛠 🧪 win](https://ews-build.webkit.org/#/builders/10/builds/172070 "An unexpected error occured. Recent messages:Pull request contains relevant changes; Failed to print configuration") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/76/builds/12719 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/85/builds/2615 "Built successfully") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/36/builds/94857 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/12/builds/109557 "Built successfully") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/19/builds/108360 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/78/builds/9751 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/3/builds/92969 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/35/builds/37408 "Passed tests") | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/9/builds/91609 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/73/builds/24473 "Passed tests") | [❌ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/34/builds/79157 "Found 2 new API test failures: /WebKitGTK/TestWebKitAccessibility:/webkit/WebKitAccessibility/accessible/state-changed, /WebKitGTK/TestWebKitAccessibility:/webkit/WebKitAccessibility/accessible/event-listener (failure)") | 
| [❌ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/1/builds/92822 "Found 1 new JSC stress test failure: wasm.yaml/wasm/lowExecutableMemory/exports-oom.js.default-wasm (failure)") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/81/builds/5170 "Built successfully") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/70/builds/25892 "Passed tests") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/38/builds/89144 "Built successfully") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/83/builds/2849 "Built successfully and passed tests") | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/82/builds/5330 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/84/builds/2341 "Passed tests") | [✅ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/46/builds/29626 "Passed tests") | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/80/builds/11350 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/62/builds/45381 "Passed tests") | [✅ 🛠 jsc-mips](https://ews-build.webkit.org/#/builders/37/builds/92070 "Built successfully") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/74/builds/5937 "Built successfully and passed tests") | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/79/builds/7062 "Built successfully") | | [✅ 🧪 jsc-mips-tests](https://ews-build.webkit.org/#/builders/45/builds/20627 "Passed tests") | 
| | | | | 
<!--EWS-Status-Bubble-End-->

mwyrzykowski commented on Source/WebGPU/WebGPU/BindGroupLayout.mm:
```diff
@@ -109,21 +109,28 @@ static bool isPresent(const WGPUStorageTextureBindingLayout& storageTexture)
 
     UNUSED_PARAM(storageTexture);
-    auto descriptor = [MTLArgumentDescriptor new];
+    auto descriptor = [MTLArgumentDescriptor argumentDescriptor];
```
This change isn't needed, it could be reverted.

djg commented on Source/WebGPU/WebGPU/BindGroupLayout.h:
```diff
@@ -41,13 +41,9 @@ class Device;
     WTF_MAKE_FAST_ALLOCATED;
 public:
-    static Ref<BindGroupLayout> create(id<MTLArgumentEncoder> vertexArgumentEncoder, id<MTLArgumentEncoder> fragmentArgumentEncoder, id<MTLArgumentEncoder> computeArgumentEncoder)
+    static Ref<BindGroupLayout> create(HashMap<uint32_t, WGPUShaderStageFlags>&& stageMapTable, id<MTLArgumentEncoder> vertexArgumentEncoder = nil, id<MTLArgumentEncoder> fragmentArgumentEncoder = nil, id<MTLArgumentEncoder> computeArgumentEncoder = nil)
```
Why not pass as an array of `id<MTLArgumentEncoder>`s instead of individually? (And instead of `m_vertexArgumentEncoder`, access would be via `m_argumentEncoder[vertex]`)

mwyrzykowski replied:
It is possible but I would need to make `vertex` / `fragment` / `compute` accessible via the BindGroupLayout too, currently they are only defined in BindGroup.mm

litherum commented on Source/WebGPU/WebGPU/BindGroup.mm:
```diff
@@ -50,127 +50,27 @@ static bool textureViewIsPresent(const WGPUBindGroupEntry& entry)
+    uint32_t entryCount = descriptor.entryCount;
+    const auto& bindGroupLayout = WebGPU::fromAPI(descriptor.layout);
     Vector<BindableResource> resources;
-#if HAVE(TIER2_ARGUMENT_BUFFERS)
```
Wow, so much red!

litherum commented on Source/WebGPU/WebGPU/BindGroup.mm:
```diff
@@ -50,127 +50,27 @@ static bool textureViewIsPresent(const WGPUBindGroupEntry& entry)
     if (descriptor.nextInChain)
         return BindGroup::createInvalid(*this);
 
+    constexpr WGPUShaderStage shaderStage[] = { WGPUShaderStage_Vertex, WGPUShaderStage_Fragment, WGPUShaderStage_Compute };
```
I'd like your opinion about something. On a previous review, Kimmo suggested considering the `WGPU` types as external types, only to be used at the API boundary of the framework. Just like how we have `WGPUBuffer` as an API type, but internally we have `WebGPU::Buffer`, we could extend that relationship to _all_ `WGPU` types. The "binding" code to turn all `WGPU` types into their corresponding `WebGPU::` types would be generated code, driven by IDL as a custom build step.

What do you think about the tradeoff of having cleaner internal/external boundaries, vs the added infrastructure (and potential runtime cost?) of transforming a bunch of types into different types that have the same expressivity?

mwyrzykowski replied:
I think this makes a lot of sense since if the external (WGPU) types change in any form (renames or values change), we do not want to have a lot of compilation or runtime breakage throughout our code.

mwyrzykowski replied:
I'll make this change prior to merging

litherum commented on Source/WebGPU/WebGPU/BindGroup.mm:
```diff
@@ -50,127 +50,27 @@ static bool textureViewIsPresent(const WGPUBindGroupEntry& entry)
 
-        return BindGroup::create(argumentBuffer[0], argumentBuffer[1], argumentBuffer[2], WTFMove(resources), *this);
+    id<MTLArgumentEncoder> argumentEncoder[] = { bindGroupLayout.vertexArgumentEncoder(), bindGroupLayout.fragmentArgumentEncoder(), bindGroupLayout.computeArgumentEncoder() };
+    for (size_t i = 0; i < shaderStageLength; ++i) {
```
How do you feel about naming these values of `i` in an enum, and doing this in a lambda instead of a loop? `EnumeratedArray` is a pretty elegant thing to use for situations like these.

```
enum class ShaderType {
    Vertex,
    Fragment,
    Compute
};
EnumeratedArray<ShaderType, id<MTLArgumentEncoder>, ShaderType::Compute> argumentEncoders;
auto foo = [](ShaderType type) {
    ... argumentEncoders[type] ...
}
foo(ShaderType::Vertex);
foo(ShaderType::Fragment);
foo(ShaderType::Compute);
```

mwyrzykowski replied:
I think using EnumeratedArray makes sense, I'm attempting to apply the suggestion of using EnumeratedArray  without using the lambdas since it doubles the code lines of the first for loop which I think impacts readability.

litherum commented on Source/WebGPU/WebGPU/BindGroup.mm:
```diff
@@ -179,31 +79,35 @@ static auto sizeOfEntries(const WGPUBindGroupDescriptor& descriptor, BindGroupLa
-            ASSERT_NOT_REACHED();
-            return BindGroup::createInvalid(*this);
+        auto stages = bindGroupLayout.stagesForBinding(entry.binding);
+        for (size_t currentStage = 0; currentStage < shaderStageLength; ++currentStage) {
```
(ditto, if you decide that it's a good idea)

litherum commented on Source/WebGPU/WebGPU/BindGroup.mm:
```diff
@@ -179,31 +79,35 @@ static auto sizeOfEntries(const WGPUBindGroupDescriptor& descriptor, BindGroupLa
+                id<MTLTexture> texture = WebGPU::fromAPI(entry.textureView).texture();
+                [argumentEncoder[currentStage] setTexture:texture atIndex:index++];
+                resources.append({ texture, MTLResourceUsageRead, renderStage });
+            }
```
This looks pretty nice

litherum commented on Source/WebGPU/WebGPU/BindGroup.mm:
```diff
@@ -179,31 +79,35 @@ static auto sizeOfEntries(const WGPUBindGroupDescriptor& descriptor, BindGroupLa
     }
 
-    return BindGroup::create(vertexArgumentBuffer, fragmentArgumentBuffer, computeArgumentBuffer, WTFMove(resources), *this);
+    constexpr int vertex = 0, fragment = 1, compute = 2;
```
Right, the enum would save you from having to do this

litherum commented on Source/WebGPU/WebGPU/RenderPipeline.mm:
```diff
@@ -488,6 +488,32 @@ static MTLVertexStepFunction stepFunction(WGPUVertexStepMode stepMode)
 
 RenderPipeline::~RenderPipeline() = default;
 
+#if HAVE(METAL_BUFFER_BINDING_REFLECTION)
```
Alright

mwyrzykowski committed:
[WebGPU] Support tier1 level argument buffers
https://bugs.webkit.org/show_bug.cgi?id=250123
<radar://103905011>

Reviewed by Myles C. Maxfield.

Support Tier1 argument buffers which are limited to
pre-A13 devices and some Intel Macs.

Tested on a pre-A13 iPad Pro.

* Source/WebGPU/WebGPU/BindGroup.mm:
(WebGPU::sizeOfEntries):
(WebGPU::Device::createBindGroup):
Support tier1 ABs.

* Source/WebGPU/WebGPU/BindGroupLayout.h:
(WebGPU::BindGroupLayout::create):
Simplify create function.

* Source/WebGPU/WebGPU/BindGroupLayout.mm:
(WebGPU::createArgumentDescriptor):
(WebGPU::addDescriptor):
(WebGPU::Device::createBindGroupLayout):
(WebGPU::BindGroupLayout::BindGroupLayout):
(WebGPU::BindGroupLayout::encodedLength const):
Support tier1 ABs.

* Source/WebGPU/WebGPU/RenderPipeline.mm:
(WebGPU::createEntryFromStructMember):
(WebGPU::RenderPipeline::getBindGroupLayout):
Call Device::createBindLayout instead and pass some additional
information to be compatible with Tier1 ABs.

Canonical link: https://commits.webkit.org/258651@main
